CondCore/CondDB/src/IOVProxy.cc

```diff
@@ -292,6 +300,135 @@ namespace cond {
+      for (int i = 0; i < reco::BeamSpot::dimension; ++i) {
+        for (int j = 0; j < reco::BeamSpot::dimension; ++j) {
+#ifdef BEAM_SPOT_DIAGNOSTICS_DEBUG
+          switch (testCaseNr) {
+            case 0:
+              matrix(i, j) = beamspotObj->covariance(i, j);
+              break;
+            case 1:
+              matrix(i, j) = 0;
+              break;
+            case 2:
+              if (i < 3 && j < 3)
+                matrix(i, j) = invalidMatrix[i][j];
+              break;
+          }
+#else
+          matrix(i, j) = beamspotObj->covariance(i, j);
+#endif
+        }
+      }
+
+#ifdef BEAM_SPOT_DIAGNOSTICS_DEBUG
+      if (testCaseNr == 3) {
+        return reco::BeamSpot();
+      }
+      testCaseNr++;
+#endif
+      reco::BeamSpot beamSpot = reco::BeamSpot(apoint,  // (force formatting)
+                                               beamspotObj->sigmaZ(),
+                                               beamspotObj->dxdz(),
+                                               beamspotObj->dydz(),
+                                               beamspotObj->beamWidthX(),
+                                               matrix);
+      beamSpot.setBeamWidthY(beamspotObj->beamWidthY());
+      beamSpot.setEmittanceX(beamspotObj->emittanceX());
+      beamSpot.setEmittanceY(beamspotObj->emittanceY());
+      beamSpot.setbetaStar(beamspotObj->betaStar());
+      return beamSpot;
+    }
+
+    bool printBeamSpotDiagnostics(std::shared_ptr<IOVProxyData> iovProxyData,
+                                  std::shared_ptr<SessionImpl> sessionImpl,
+                                  cond::Time_t lowerGroup,
+                                  cond::Time_t higherGroup) {
+      bool printedDiagnostics = false;
+      if (iovProxyData->tagInfo.payloadType == "BeamSpotOnlineObjects") {
+        cond::persistency::Session session(sessionImpl);
+        session.transaction().start(true);
+#ifdef BEAM_SPOT_DIAGNOSTICS_DEBUG
+        int iSeq = 0;
+#endif
+        for (const auto& [iov, hash] : iovProxyData->iovSequence) {
+          std::unique_ptr<BeamSpotOnlineObjects> beamspotObj = session.fetchPayload<BeamSpotOnlineObjects>(hash);
+          reco::BeamSpot beamSpot = beamSpotFromBSOnlineObjects(beamspotObj);
+
+          //Check validity of BeamSpot by validating error of VertexState created from it
+          VertexState beamVertexState(beamSpot);
+          bool vertexFromBSInvalid = (beamVertexState.error().cxx() <= 0.) || (beamVertexState.error().cyy() <= 0.) ||
+                                     (beamVertexState.error().czz() <= 0.);
+
+          //Check validity of BeamSpot by confirming it can be inverted
+          int inversionFail = 2;  //the .Inverse(...) will set it to 0 or 1
+          reco::BeamSpot::CovarianceMatrix errorInverse = beamSpot.covariance().Inverse(inversionFail);
+
+          double det;  //TODO do we check BS validity by det?
+          beamSpot.covariance().Det2(det);
```

Comment:
Is that a good way to check the validity of beamSpotObjects? Wouldn't that be too big of an overhead for HLT? What would you do differently?